File: Engine/UI/LayerWidget.py
# File name: LayerWidget.py
# This file is part of: pyengine
#
# LICENSE
#
# The contents of this file are subject to the Mozilla Public License
# Version 1.1 (the "License"); you may not use this file except in
# compliance with the License. You may obtain a copy of the License at
# http://www.mozilla.org/MPL/
#
# Software distributed under the License is distributed on an "AS IS"
# basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See
# the License for the specific language governing rights and limitations
# under the License.
#
# Alternatively, the contents of this file may be used under the terms
# of the GNU General Public license (the  "GPL License"), in which case
# the provisions of GPL License are applicable instead of those above.
#
# FEEDBACK & QUESTIONS
#
# For feedback and questions about pyengine please e-mail one of the
# authors named in the AUTHORS file.
########################################################################
from __future__ import unicode_literals, print_function, division
from our_future import *

__all__ = ["DesktopLayer", "WindowLayer", "PopupLayer"]

import CSS.Minilanguage

from WidgetBase import ParentWidget
from ScreenWidget import ScreenWidget

class LayerWidget(ParentWidget):
    def hit_test(self, p):
        self.realign()
        if not p in self.AbsoluteRect:
            return None
        return self._hit_test(p)

    def hit_test_with_chain(self, p):
        self.realign()
        if not p in self.AbsoluteRect:
            return False
        return self._hit_test_with_chain(p)

class DesktopLayer(LayerWidget):
    def do_align(self):
        for child in self:
            child.AbsoluteRect = self.AbsoluteRect

class WindowLayer(LayerWidget):
    def __init__(self, parent, **kwargs):
        super(WindowLayer, self).__init__(parent, **kwargs)

    def do_align(self):
        rect = self.AbsoluteRect
        for item in self:
            if isinstance(item, ModalWindowLayer):
                item.AbsoluteRect = rect

class PopupLayer(LayerWidget):
    def __init__(self, parent, **kwargs):
        super(PopupLayer, self).__init__(parent, **kwargs)
        self._current_root_menu = None

    def hit_test(self, p):
        if self._current_root_menu:
            hit = ParentWidget.hit_test(self, p)
            if hit is self and p in self._current_root_menu.AbsoluteRect:
                return self._current_root_menu.hit_test(p)
            return hit
        else:
            return super(PopupLayer, self).hit_test(p)

    def on_mouse_click(self, x, y, button, modifiers, nth):
        if self._current_root_menu:
            self._current_root_menu.hide_submenu()
            self._current_root_menu = None

    @property
    def CurrentRootMenu(self):
        return self._current_root_menu

    @CurrentRootMenu.setter
    def CurrentRootMenu(self, value):
        self._current_root_menu = value

class ModalWindowLayer(ParentWidget):
    # opaque layer widget which eats all other events
    def __init__(self, parent, window, **kwargs):
        super(ModalWindowLayer, self).__init__(parent, **kwargs)
        self._window = window
        window._on_close_hook = self._handle_window_close

        self.add(window)
        if self.Parent:
            self.RootWidget.focus(self._window)

    def _parent_changed(self):
        super(ModalWindowLayer, self)._parent_changed()
        if self.Parent:
            self.AbsoluteRect = self.Parent.AbsoluteRect
            self.invalidate()

    def _handle_window_close(self, sender):
        self.invalidate()
        self.Parent.remove(self)

CSS.Minilanguage.ElementNames().register_widget_class(DesktopLayer)
CSS.Minilanguage.ElementNames().register_widget_class(WindowLayer)
CSS.Minilanguage.ElementNames().register_widget_class(PopupLayer)
CSS.Minilanguage.ElementNames().register_widget_class(ModalWindowLayer, "ModalLayer")

File: CEngine/UI/Widgets/LayerWidget.cpp
/**********************************************************************
File name: LayerWidget.cpp
This file is part of: Pythonic Engine

LICENSE

The contents of this file are subject to the Mozilla Public License
Version 1.1 (the "License"); you may not use this file except in
compliance with the License. You may obtain a copy of the License at
http://www.mozilla.org/MPL/

Software distributed under the License is distributed on an "AS IS"
basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
License for the specific language governing rights and limitations under
the License.

Alternatively, the contents of this file may be used under the terms of
the GNU General Public license (the  "GPL License"), in which case  the
provisions of GPL License are applicable instead of those above.

FEEDBACK & QUESTIONS

For feedback and questions about pyengine please e-mail one of the
authors named in the AUTHORS file.
**********************************************************************/
#include "LayerWidget.hpp"

#include "RootWidget.hpp"
#include "WindowWidget.hpp"

namespace PyEngine { namespace UI {

/* PyEngine::UI::LayerWidget */

/* PyEngine::UI::DesktopLayer */

void DesktopLayer::do_align()
{
    const Rect r = absolute_rect();
    for (auto& child: *this) {
        Margin m = child->computed_style().margin();
        coord_int_t w, h;
        std::tie(w, h) = child->get_dimensions();
        Rect obj = Rect(r.get_left(), r.get_top(), w, h);
        m.deautoify(obj, r);
        obj = r;
        obj.shrink(m);
        child->absolute_rect() = obj;
        child->invalidate_alignment();
    }
}

bool DesktopLayer::is_element(const std::string &name) const
{
    return (name == "desktoplayer");
}

/* PyEngine::UI::WindowLayer */

void WindowLayer::do_align()
{
    for (auto& child: *this) {
        ModalWindowLayer *modal = dynamic_cast<ModalWindowLayer*>(child);
        if (modal) {
            modal->absolute_rect() = absolute_rect();
        }
    }
}

bool WindowLayer::is_element(const std::string &name) const
{
    return (name == "windowlayer");
}

/* PyEngine::UI::PopupLayer */

WidgetPtr PopupLayer::hittest(const Point &p)
{
    if (_current_root_menu) {
        WidgetPtr hit = LayerWidget::hittest(p);
        if ((hit == this) && (_current_root_menu->absolute_rect().contains(p)))
        {
            return _current_root_menu->hittest(p);
        }
        return hit;
    } else {
        return LayerWidget::hittest(p);
    }
}

bool PopupLayer::hittest_with_chain(const Point &p, HitChain &chain)
{
    if (!_hittest(p)) {
        return false;
    }

    if (_current_root_menu) {
        _current_root_menu->hittest_with_chain(p, chain);
    }

    chain.push_back(this);
    return true;
}

bool PopupLayer::is_element(const std::string &name) const
{
    return (name == "popuplayer");
}

bool PopupLayer::ev_mouse_click(
    int x, int y,
    MouseButton button,
    KeyModifiers modifiers,
    unsigned int nth)
{
    if (_current_root_menu) {
        /* TODO: close menu */
    }
    return false;
}

/* PyEngine::UI::ModalWindowLayer */

ModalWindowLayer::ModalWindowLayer(
        ModalWindowLayer::WindowWidgetPtr window):
    _window(window),
    _close_hook(_window->on_close_hook().connect(
        sigc::mem_fun(this, &ModalWindowLayer::_handle_window_close)))
{
    add(_window);
    RootPtr root = get_root();
    if (root) {
        root->focus(_window);
    }
}

ModalWindowLayer::~ModalWindowLayer()
{
    _close_hook.disconnect();
}

void ModalWindowLayer::_parent_changed()
{
    ParentWidget::_parent_changed();
    WidgetPtr parent = get_parent();
    if (parent) {
        absolute_rect() = parent->absolute_rect();
        invalidate();
    }
    RootPtr root = get_root();
    if (root && _window) {
        root->focus(_window);
    }
}

void ModalWindowLayer::_handle_window_close(WidgetPtr sender)
{
    invalidate();
    _close_hook.disconnect();
    ParentPtr parent = get_parent();
    if (parent) {
        parent->remove(this);
    }
}

bool ModalWindowLayer::is_element(const std::string &name) const
{
    return (name == "modalwindowlayer");
}

}
}
